### PR TITLE
Cosine bf

### PR DESCRIPTION
**Pull request by jbpoline**
It looks like the cosine_drift was not giving exactly the answer that spm does, while this implementation does. I also found that the basis (excluding the non - normalized constant term) was not exactly othonormal, this one is. I added a test using the output of spm_dctmtx and corrected the corresponding tests.

### CODE REVIEW

File: nipy/modalities/fmri/design_matrix.py [hunk 1]
```diff
@@ -85,6 +85,42 @@ def _cosine_drift(hfcut, frametimes):
...
+    hfcut = 1./hfcut # input parameter is the period - originally the freq.
+
+    dt,T = tim[1]-tim[0], tim.max() # == (N-1)*dt    
+    order = int(np.floor(2*N*hfcut*dt)) # s.t. hfcut = 1/(2*dt) yields N
+    # print order
+    cdrift = np.zeros((N, order))
+    nfct = np.sqrt(2.0/N)
+    
+    for k in range(1,order):
+        #  nfct * cos(pi*k*(2n+1)/(2*N))  n:[0,N-1], k:[0,N-1]
```
Comment (bthirion) [inline nipy/modalities/fmri/design_matrix.py]: Please remove this line

File: nipy/modalities/fmri/design_matrix.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ def _poly_drift(order, frametimes):
     return pol
 
 
-def _cosine_drift(hfcut, frametimes):
+def _cosine_drift_(hfcut, frametimes):
```
Comment (bthirion) [inline nipy/modalities/fmri/design_matrix.py]: if this function is no longer used, please remove it.

**Comment by bthirion** [comment]
Thx for taking care of this.

**Comment by jbpoline** [comment]
Thanks - sorry I didn't see earlier that you had already responded, missed the notification.

File: nipy/modalities/fmri/design_matrix.py [hunk 3]
```diff
@@ -61,30 +61,42 @@ def _poly_drift(order, frametimes):
...
     """
-    tmax = float(frametimes.max())
-    tsteps = len(frametimes)
-    order = int(np.floor(2 * float(tmax) / float(hfcut)) + 1)
-    cdrift = np.zeros((tsteps, order))
-    for k in range(1, order):
-        cdrift[:, k - 1] = np.sqrt(2.0 / tmax) * np.cos(
-            np.pi * (frametimes / tmax + 0.5 / tsteps) * k)
-    cdrift[:, order - 1] = np.ones_like(frametimes)
+    N = len(tim)
```
Comment (bthirion) [inline nipy/modalities/fmri/design_matrix.py]: Capital letters should be used for constants. Rather use e.g. `len_sampling_times`

File: nipy/modalities/fmri/design_matrix.py [hunk 4]
```diff
@@ -61,30 +61,42 @@ def _poly_drift(order, frametimes):
...
+    n_tim = np.arange(N)
+    hfcut = 1./hfcut # input parameter is the period - originally the freq.
```
Comment (bthirion) [inline nipy/modalities/fmri/design_matrix.py]: hfcut = 1. / hfcut

**Comment by jbpoline** [comment]
I added the spm_dcmtx txt file data to the npz file. Let me know if you find anything else to be mended.

File: nipy/modalities/fmri/design_matrix.py [hunk 5]
```diff
@@ -61,30 +61,42 @@ def _poly_drift(order, frametimes):
     return pol
 
 
-def _cosine_drift(hfcut, frametimes):
-    """Create a cosine drift matrix
+def _cosine_drift(period_cut, tim):
```
Comment (matthew-brett) [inline nipy/modalities/fmri/design_matrix.py]: `frametimes` easier to understand than `tim`

File: nipy/modalities/fmri/design_matrix.py [hunk 6]
```diff
@@ -61,30 +61,42 @@ def _poly_drift(order, frametimes):
...
-    tsteps = len(frametimes)
-    order = int(np.floor(2 * float(tmax) / float(hfcut)) + 1)
-    cdrift = np.zeros((tsteps, order))
+    len_tim = len(tim)
+    n_times = np.arange(len_tim)
+    hfcut = 1./ period_cut # input parameter is the period  
+
+    dt,T = tim[1]-tim[0], tim.max() # == (len_tim-1)*dt    
+    order = int(np.floor(2*len_tim*hfcut*dt)) # s.t. hfcut = 1/(2*dt) yields len_tim
+    # print order
```
Comment (matthew-brett) [inline nipy/modalities/fmri/design_matrix.py]: Debug cruft?

File: nipy/modalities/fmri/design_matrix.py [hunk 7]
```diff
@@ -61,30 +61,42 @@ def _poly_drift(order, frametimes):
...
+    cdrift = np.zeros((len_tim, order))
+    nfct = np.sqrt(2.0/len_tim)
+    
     for k in range(1, order):
-        cdrift[:, k - 1] = np.sqrt(2.0 / tmax) * np.cos(
-            np.pi * (frametimes / tmax + 0.5 / tsteps) * k)
-    cdrift[:, order - 1] = np.ones_like(frametimes)
+        #  nfct * cos(pi*k*(2n+1)/(2*len_tim))  n:[0,len_tim-1], k:[0,len_tim-1]
```
Comment (matthew-brett) [inline nipy/modalities/fmri/design_matrix.py]: What is this comment?

File: nipy/modalities/fmri/design_matrix.py (hunk 5 above)
Comment (jbpoline) [inline nipy/modalities/fmri/design_matrix.py]: In this specific case I tried to keep short names for variable to more easily match the formula to the code, but ok  to change to longer names if people find it easier to read

**Comment by bthirion** [comment]
Thanks for the changes. I propose a couple of updates in my PR. Then it should be OK for merge.

**Comment by matthew-brett** [comment]
@jbpoline  - where are we on this one?

**Comment by bthirion** [comment]
I think it's good to go (note that it has been introduced into nistats anyway).

**Comment by matthew-brett** [comment]
Thanks JB - sorry to be slow - merging.

**Pull request merged**